taichi/transforms/auto_diff.cpp

```diff
@@ -1143,6 +1143,43 @@ class MakeDual : public ADTransform {
     }
   }
 
+  void visit(RangeForStmt *for_stmt) override {
+    std::vector<Stmt *> statements;
+    // always make a copy since the list can be modified.
+    for (auto &stmt : for_stmt->body->statements) {
+      statements.push_back(stmt.get());
+    }
+    for (auto stmt : statements) {
+      current_stmt = stmt;
+      stmt->accept(this);
+    }
+  }
+
+  void visit(StructForStmt *for_stmt) override {
+    // alloca_block = for_stmt->body.get();
+    for_stmt->body->accept(this);
+  }
+
+  void visit(LocalLoadStmt *stmt) override {
+    // TI_ASSERT(!needs_grad(stmt->ret_type));
+    accumulate(stmt, dual(stmt->src.data[0].var));
+  }
+
+  void visit(LocalStoreStmt *stmt) override {
+    // Clear the dual of the dest before local store,
+    // Because LocalStoreStmt overwrites the dest,
+    // If the alloca serves as the dest of multiple LocalStoreStmt, only the
+    // last LocalStoreStmt should be taken account of, i.e, its history should
+    // be cleared
+    if (needs_grad(stmt->dest->ret_type)) {
```

Comment:
do we still need unified `needs_grad` interface?

Reply:
I think this `needs_grad` is used to prevent creating `alloca` of `alloca`?

Reply:
Ah I think I misunderstood the `needs_grad` here (I was thinking maybe it's more proper to have `needs_dual`, then realized it only means `is_real(dt)`....

Reply:
It’s indeed confusing here. IOI